### DISCUSSION TITLE
Struggling to parse GML in RDF+XML files

### DISCUSSION THREAD

**Discussion by dodinh** [apache/jena#2667] [Q&A] [answered]
Hi everyone,

I have RDF files in XML format that contain GML snippets like this:

```
<dct:spatial>
    <dct:Location>
        <dcat:bbox {MYSTERY}>
            <gml:Polygon>
                <gml:exterior>
                    <gml:LinearRing>
                        <gml:posList>2.293197561936239 48.85825079553791 2.294494006792206 48.85739338328426 2.295803956283663 48.858277450605584 2.294500759109951 48.85910819310513 2.293197561936239 48.85825079553791</gml:posList>
                    </gml:LinearRing>
                </gml:exterior>
            </gml:Polygon>
        </dcat:bbox>
    </dct:Location>
</dct:spatial>
```

As they come from various sources and in various degrees of well-formedness, I have to preprocess them anyway. However, I struggle with the {MYSTERY} part. I tried different combinations:

1. {MYSTERY} = `rdf:parseType="Literal"`
    this expectedly lets me handle the GML as a literal but the datatype is set as *http://www.w3.org/1999/02/22-rdf-syntax-ns#XMLLiteral* which obviously `GeometryWrapper` cannot handle
2. {MYSTERY} = `rdf:datatype="http://www.opengis.net/ont/geosparql#gmlLiteral"`
    this delivers a Resource instead of a Literal
3. {MYSTERY} = `rdf:parseType="Literal" rdf:datatype="http://www.opengis.net/ont/geosparql#gmlLiteral"`
    this looked like it should work, but the datatype is ignored and set to  *http://www.w3.org/1999/02/22-rdf-syntax-ns#XMLLiteral* like in 1.

My current workaround is explicitly setting the datatype in the code via
```
GeometryWrapper wrapper = GeometryWrapper.extract(geometryNode.asLiteral().getString(), "http://www.opengis.net/ont/geosparql#gmlLiteral");
```

As I'm expecting different datatypes here (GeoJSON, WKT), this workaround is possible but tedious, seems inelegant and easily broken. Is there another solution?

I'm using Apache Jena 5.1.0

**Comment by afs** [comment, marked as answer]
Hi @dodinh,

Q1. `rdf:parseType="Literal"` 

RDF/XML doesn't have a way to say "parse as structured XML and set the datatype to this URI".  
You may also notice the lexical form of the RDF literal produced includes a `xmlns:gml=` declaration.
This is part of RDF/XML parsing to produce a valid `rdf:XMLLiteral`. To be valid, it must have must be a standalone XML fragement and have namespace declarations.

Q2. `rdf:datatype="http://www.opengis.net/ont/geosparql#gmlLiteral"` 

The body of the property which has `rdf:datatype` is the lexical form for the literal. It has to be a simple string with any XML characters escaped e.g. `&lt;`. Structured XML is not accepted.

Q3. `rdf:parseType=`  `rdf:datatype=`

That's a bug, it should be an error. I've raised an issue for this #2669. In Jena4 it correctly causes a parse error.

So the two choices are:

1. `rdf:datatype=` with an escaped XML fragment
2. Parse as `rdf:parseType=` and convert. 

The first means the input file needs changing which is not ideal.

The second way is to feed the parser output into a `StreamRDF` by calling the parser via `RDFParser.source(...).parse(myStreamRDF)` and have `myStreamRDF` convert the datatype to GML.

**Reply by dodinh** [reply]
Thank you very much for the quick and exhaustive answer and explanations, and for the issue/fix for Q3.
